src/main/kotlin/org/rust/ide/annotator/RsMacroExpansionHighlightingPass.kt

```diff
@@ -0,0 +1,153 @@
+import com.intellij.lang.annotation.AnnotationSession
+import com.intellij.lang.annotation.Annotator
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import org.rust.ide.utils.isEnabledByCfg
+import org.rust.lang.core.macros.*
+import org.rust.lang.core.psi.RsMacroCall
+import org.rust.lang.core.psi.ext.descendantsOfType
+import org.rust.lang.core.psi.ext.expansion
+import org.rust.stdext.removeLast
+
+class RsMacroExpansionHighlightingPassFactory(
+    val project: Project,
+    registrar: TextEditorHighlightingPassRegistrar
+) : DirtyScopeTrackingHighlightingPassFactory {
+    private val myPassId: Int = registrar.registerTextEditorHighlightingPass(
+        this,
+        null,
+        null,
+        false,
+        -1
+    )
+
+    override fun createHighlightingPass(file: PsiFile, editor: Editor): TextEditorHighlightingPass? {
+        if (project.macroExpansionManager.macroExpansionMode !is MacroExpansionMode.New) return null
+        val restrictedRange = FileStatusMap.getDirtyTextRange(editor, passId) ?: return null
+        return RsMacroExpansionHighlightingPass(file, restrictedRange, editor.document)
+    }
+
+    override fun getPassId(): Int = myPassId
+}
+
+class RsMacroExpansionHighlightingPass(
+    private val file: PsiFile,
+    private val restrictedRange: TextRange,
+    document: Document
+) : TextEditorHighlightingPass(file.project, document) {
+    private val results = mutableListOf<HighlightInfo>()
+
+    private fun createAnnotators(): List<Annotator> = listOf(
+        RsEdition2018KeywordsAnnotator(),
+        RsHighlightingAnnotator(),
+        RsHighlightingMutableAnnotator()
+    )
+
+    override fun doCollectInformation(progress: ProgressIndicator) {
+        val macros = mutableListOf<PreparedMacroCall>()
+        for (macro in file.descendantsOfType<RsMacroCall>()) {
+            if (macro.macroArgument?.textRange?.intersects(restrictedRange) != true) continue
+            macros += macro.prepare() ?: continue
+        }
+        if (macros.isEmpty()) return
+
+        val annotators = createAnnotators()
+        val holder = AnnotationHolderImpl(AnnotationSession(file))
+        while (macros.isNotEmpty()) {
+            val macro = macros.removeLast()
+            for (element in macro.elementsForHighlighting) {
+                for (ann in annotators) {
+                    ProgressManager.checkCanceled()
+                    ann.annotate(element, holder)
+                }
+                if (element is RsMacroCall) {
+                    macros += element.prepare() ?: continue
+                }
+            }
+            mapAndCollectAnnotations(macro, holder)
```

Comment:
It seems we can improve performance from `O(annotations * ranges)` to `O(annotations * log(annotations) + ranges * log(ranges) + annotations + ranges)` if we sort annotations and expansion ranges and when make linear pass through two lists
But it would be great to measure somehow how mich time current implementation takes and decide if we need to rewrite our naive mapping